Avoid unnecessary use of TCS

jdom opened the pull request:
Removes the unnecessary wrapping of a `TaskCompletionSource` into another `TaskCompletionSource`.

For some reason 1 test (`Generic_SimpleGrainControlFlow`) started deadlocking after the change (and ending in a `TimeoutException` after 30 seconds). The test was doing the wrong thing in the first place (blocking on `.Result`), nevertheless because the `OutsideRuntimeClient` executes the response callbacks in the same thread as where it is doing the message pump, then the `.Result` resulted in the message pump not running. I'm not sure why this was not the case before the change, so I'm worried that this could have breaking changes for people that were also doing blocking waits (albeit wrong).
This could be fixed by executing the callback inside a `Task.Run()` call, but that could bring bad perf implications for the people writing the code correctly and not blocking (although not sure if this path only occurs in unit tests or in an actual Orleans client, as that could be dangerous too if you are for example doing CPU intensive operations with the response, such as deserializing complex objects).

This PR is mainly a talking point, but it would be fine not to commit to master if the deadlocking I started seeing before my 2nd commit is a valid concern.

dnfclas commented:
Hi __@jdom__, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla2.dotnetfoundation.org.
    </span>

TTYL, DNFBOT;

dnfclas commented:
__@jdom__, Thanks for signing the contribution license agreement so quickly! Actual humans will now validate the agreement and then evaluate the PR.
<br />Thanks, DNFBOT;

gabikliot commented on src/Tester/GenericGrainTests.cs:
```diff
@@ -299,7 +299,7 @@ public static void MyClassCleanup()
             await grain.SetB(b);
 
             Task<string> stringPromise = grain.GetAxB();
-            Assert.AreEqual(expected, stringPromise.Result);
```
We definitely do NOT want to change this test.
This test was the only test in our CI that broke due to the wrong usage of `ExecuteSynchronosuly` in https://github.com/dotnet/orleans/pull/853.

There is nothing wrong or illegal about mixing up `await` and `.Wait` **on the client**. Indeed, it is a bad style, and people should try to avoid it (either all `await` or all blocking Wait), but TPL does allow it, and Orleans cannot prevent its customers from doing so **on the client** and thus we should support it and also test it.

jdom replied:
Then, as mentioned in the description, we should probably not rely on the TCS to sometimes avoid the deadlock, a more appropriate fix for that is probably to have the message pump to make sure to run the callbacks in a different thread that is pumping messages, otherwise it is very prone not just to deadlocks, but to prevent the message pumping to continue... If you want, I can submit another PR so that you evaluate what I think should be done in the OutsideRuntimeClient

gabikliot commented on src/Orleans/Runtime/GrainReference.cs:
```diff
@@ -339,8 +339,8 @@ protected void InvokeOneWayMethod(int methodId, object[] arguments, InvokeMethod
 
-            resultTask = OrleansTaskExtentions.ConvertTaskViaTcs(resultTask);
-            return (T) await resultTask;
+            object result = await resultTask.ConfigureAwait(false);
```
@jdom , why `ConfigureAwait(false);`?
This code runs both on the client and on the silo, and we don't do `ConfigureAwait(false);` on the silo, ever.

What is it good for? (I mean, I know in general what ConfigureAwait(false);  is good for. Why did you put it here?)

gabikliot commented:
Indeed, it looks like `ConvertTaskViaTcs` in this particular place is totally unnecessary and can be simply replaced with `await` (I don't understand why ` ConfigureAwait(false)` is needed).

The reason it was added in the first place is, as far as I remember now, since it was some years ago, is the following:
with simple `await` there was this one case that caused Orleans client to deadlock. When Orleans client runs on a UI thread, such as the case in out Web sample (https://github.com/dotnet/orleans/blob/master/Samples/AzureWebSample/WebRole/Default.aspx.cs), it used to deadlock, on  `AzureClient.Initialize`, since  `AzureClient.Initialize` is making a blocking TPL call internally. It had something to do with `SynchronizationContext` in UI threads.
The `ConvertTaskViaTcs` solved that issue back then.

Therefore, my recommendation is to test at least with our Azure Web sample, and also remember that place for future, whenever people are complaining about non responsive Orleans in UI apps.

More context:
http://www.tugberkugurlu.com/archive/the-perfect-recipe-to-shoot-yourself-in-the-foot-ending-up-with-a-deadlock-using-the-c-sharp-5-0-asynchronous-language-features
 
http://channel9.msdn.com/Events/BUILD/BUILD2011/TOOL-829T
 
http://stackoverflow.com/questions/13489065/best-practice-to-call-configureawait-for-all-server-side-code
 
“When you are awaiting on a method with await keyword, compiler generates bunch of code in behalf of you. One of the purposes of this action is to handle synchronization with the UI (or main) thread. The key component of this feature is the SynchronizationContext.Current which gets the synchronization context for the current thread. SynchronizationContext.Current is populated depending on the environment you are in. The GetAwaiter method of Task looks up for SynchronizationContext.Current. If current synchronization context is not null, the continuation that gets passed to that awaiter will get posted back to that synchronization context."

"When consuming a method, which uses the new asynchronous language features, in a blocking fashion, you will end up with a deadlock if you have an available SynchronizationContext. When you are consuming such methods in a blocking fashion (waiting on the Task with Wait method or taking the result directly from the Result property of the Task), you will block the main thread at the same time. When eventually the Task completes inside that method in the threadpool, it is going to invoke the continuation to post back to the main thread because SynchronizationContext.Current is available and captured. But there is a problem here: the UI thread is blocked and you have a deadlock!”

veikkoeeva commented:
@gabikliot 
> This code runs both on the client and on the silo, and we don't do ConfigureAwait(false); on the silo, ever.

Hmm, there are some places in the relational storage code where it is used. Firstly it is used with ADO.NET in [RelationalStorage](https://github.com/dotnet/orleans/blob/029d5c00f4c804580de44732db97f5be8d4cc405/src/Orleans/RelationalStorage/RelationalStorage.cs) and then continued in [OrleansRelationalExtensions](https://github.com/dotnet/orleans/blob/40ce3fa99a109539b8c7c71705dfa4fcdab8faa3/src/Orleans/RelationalStorage/OrleansRelationalExtensions.cs) where the results are gathered and the task is let through like [this](https://github.com/dotnet/orleans/blob/d99f81c8d043bd6070575507c795e7f76880e15b/src/Orleans/Messaging/SqlMembershipTable.cs#L120). The idea was to run the database queries in thread pool and let the original caller gather in the context it likes.

Picked from http://blog.stephencleary.com/2012/02/async-and-await.html
> A good rule of thumb is to use ConfigureAwait(false) unless you know you do need the context.

I don't know if this is a bad practice in context of Orleans, but even confusion about it might be a reason to remove the `ConfigureAwait` calls.

gabikliot commented:
@veikkoeeva , sorry, you are right. We do use `ConfigureAwait(false)` , the one place is where you pointed out. Not sure why it is required there.
We tell developer not to ever use it in the grain code, as explained here, since it escapes grain scheduler: http://dotnet.github.io/orleans/Advanced-Concepts/External-Tasks-and-Grains

As for internal runtime, there is no reason indeed to come up with a blanket statement to NEVER use it. If there is a place where it is appropriate, we should use it. But, I have not yet seen such a place. And if there is no good reason to use it, better not to , I think.

jdom commented:
Hi @gabikliot, the reason to use ConfigureAwait(false) here is because this IS a library, and at least at that point in the code there is no reason to marshal back to the synchronization context (assuming there was one, such as on a web request call that initiated this). This is exactly to avoid certain kinds of deadlocks that could unnecessarily occur. This is no different than using task.ContinueWith(...), which by default does not capture the synchronization context, and just continues in whatever thread it can.
It is totally correct to give guidance to users to never use ConfigureAwait on their APPLICATION code (unless they know what they are doing of course, but the default should be to not use it). The guidance is very different for people writing library code, and in this particular case, it is appropriate to use.

veikkoeeva commented:
@jdom I believe this gets a bit more complicated in that I remember Orleans does not set a synchronization context. I believe it was https://github.com/dotnet/orleans/issues/315 where it was discussed (by the way, if you are inclined and posses the technical chops, it'd be a great issue to pick up). So technically it wouldn't matter to set it, but having it set might blur the technical vision.

What comes to relational code, it's a micro-ORM in the vein of Dapper (though could be optimized a lot still) that has public surface and it might be handy if had `ConfigureAwait(false)`, just in case and it doesn't need the context and one might argue running the code without context is fine. There are things to consider in that what one sees when looking at other backend storage providers and if one wonders:

1. Why Table Storage code floating in Internet doesn't usually seem capture the context.
2. Why Orleans seem to capture context. Is it about turn based guarantees? Well, as @gabikliot links: http://dotnet.github.io/orleans/Advanced-Concepts/External-Tasks-and-Grains.
3. Why Relational backend doesn't capture it (see above pondering)?
4. Should one care or shoud either the Table Storage or Relational backend code change.

jdom commented:
1. Why Table Storage code floating in Internet doesn't usually seem capture the context.
Well, because it's library code. In a library you can't assume that the caller will call you with or without a synchronization context. You cannot rely on ambient stuff to be flowed. Library code does not typically need to either respond to a web request directly, nor update the UI, etc (these are cases where being in the right synchronization context is required).

	2. Why Orleans seem to capture context. Is it about turn based guarantees? Well, as @gabikliot links: http://dotnet.github.io/orleans/Advanced-Concepts/External-Tasks-and-Grains.
As an outsider and not very familiar with the Orleans codebase, I assume it's because they fixed a few places where they found that something was creating a deadlock, but did not use ConfigureAwait(false) just by omission, simplicity or maybe for lack of familiarity with it, since they could already work around the deadlocks they were seeing. Other than deadlocks, there's a potential of perf hits that are not easy to notice if for example a library (Orleans in this case) needs to wait for the synchronization context thread to be available to execute a small continuation ()maybe to just do a cast, or a small calculation) and then yield again. The code functionally still works, it's just that there will be some contention waiting for that small continuation to be re-scheduled unnecessarilly on that particular thread. I could be completely wrong with my assumption, as again, I'm not very familiar with the code nor it's evolution.

	3. Why Relational backend doesn't capture it (see above pondering)?
Similar to point 1, infrastructure code shouldn't make any assumptions about the environment it is running.

	4. Should one care or shoud either the Table Storage or Relational backend code change.
Application developers should not care. You make the call to the backend/library, they are welcome to not flow the synchronization context while THEIR code is running (and they SHOULD do this). As long as you, as the app developer, do NOT use ConfigureAwait(false), the moment that your continuation runs, then it will get back to the right synchronization context if there was any. Think of this as a hierarchy of continuations, not as a sequential chain: so it doesn't matter if somebody used ConfigureAwait(false) inside the library code you just called, you are still "saying" in your on await that you want to execute YOUR continuation (but not the intermediate ones that the library does) with the synchronization context that is captured the moment you call into the library code.

veikkoeeva commented:
@jdom I see your point about `ConfigureAwait(false)`, though I think it's not primarily about deadlocks (and for what I've seen, the guys are familiar with it). Picking up from the aforelinked Cleary one:

> Complex answer:
* If SynchronizationContext.Current is not null, then it’s the current SynchronizationContext. (UI and ASP.NET request contexts are SynchronizationContext contexts).
>
* Otherwise, it’s the current TaskScheduler (TaskScheduler.Default is the thread pool context).

I think Orleans does not set `SynchronizationContext`, so it's null. Does that mean without `ConfigureAwait(false)` the execution gets to "Orleans" and is subject to its scheduling? That may not be desirable in code that calls resources outside of the silo. If so, should all storage backends use it or is this accurate enough picture? It's clear code running in the grain context can't "leak" from the turn-based guarantees (and how context, exceptions etc. are flowing).

I wrote in Gitter I'm off to sleep now (brushed by teeth still), but this might be a good subject for discussion too!

jdom commented:
I think Orleans does not set SynchronizationContext, so it's null.
Well that's the point, especially when using OutsideRuntimeClient, Orleans can't and shouldn't make any assumptions of what the SynchronizationContext is. THEY are not setting it (but it doesn't mean it's null), but if the client is a web application or an application with UI, then THAT host will be setting the context. Orleans internally (and again, especially outside a Silo) should not rely on a sync context, so as Cleary says, it should use ConfigureAwait(false).
The ConfigureAwait I added in the GrainReference code can be run both inside or outside a silo. That part of the code (which is in fact an extremely small continuation) does not care about synchronization context nor the Orleans scheduler, so it is safe and recommended to use ConfigureAwait(false) even if it's used inside the Silo. As long as the callers of this that DO care about a potential sync context do not use ConfigureAwait(false), then we all win, as we are doing the responsible thing.

sergeybykov commented:
Should we close this one for now?

@jdom, could you also update it with information about the API (in .NET 4.6) you discovered we should be using.

jdom commented:
Yes, I think we can close this for now, as the "double-wrapping" a TCS with another TCS was actually accomplishing that the continuation executes after yielding.
There are some approaches we can take to make it cleaner (and in fact I have a few of them implemented locally and they seem to work well), but I will wait until I can spend some more time on it. 
In the long term, if we ever update to .NET 4.6, we should be creating the TCS with TaskCreationOptions.RunContinuationsAsynchronously, which acommplishes exactly that. For older framework versions we can emulate that with different alternatives (for example the ImposeAsync approach in http://stackoverflow.com/questions/22579206/how-can-i-prevent-synchronous-continuations-on-a-task ). Nevertheless, I prefer to keep it as is for now, to avoid other subtle changes that we are not aware of.

Closing this PR for now

jdom commented:
Since we updated to 4.6.1, I was going to fix this by using `TaskCreationOptions.RunContinuationsAsynchronously` when creating the first TCS, but it looks like this is no longer an issue. It was probably resolved when we added the Box an Unbox methods, which do some conversions too (didn't dig too deep on exactly when we fixed it, but added another test to make sure).
I just realized @dVakulen found the same in #2960 (just noticed that PR).

jdom commented:
@dotnet-bot test this please

jdom commented:
@dotnet-bot test this please

jdom commented:
@dotnet-bot test netfx-bvt

jdom commented:
Darn it, it looks like `TaskCreationOptions.RunContinuationsAsynchronously` is not honored when using `Task.WhenAny` and that is why `DefaultCluster.Tests.General.StatelessWorkerTests.StatelessWorkerActivationsPerSiloDoNotExceedMaxLocalWorkersCount` is failing in .NET 4.6.1 (and it's why I can't repro locally, since I have 4.6.2 installed, which I guess it's when it was fixed).
https://connect.microsoft.com/VisualStudio/feedback/details/1975944/taskcreationoptions-runcontinuationsasynchronously-not-respected-by-task-whenany-or-other-itaskcompletionaction-implementations

We could either sit on this and not merge, or alternatively use `Task.Run` (and scrap the RunContinuationsAsynchronously) before setting the result (which is effectively what we want, that user code is executed on a thread pool thread when using `OutsideRuntimeClient` (aka not within Orleans).

gabikliot commented:
It has been (at least) 2 years since this was fixed. We talked with Julian about it more than 2 years ago and TaskCreationOptions.RunContinuationsAsynchronously was already there, "in the next version of .NET".
2 years to get the code that is already implemented out to customers!

jdom commented:
Well, that's the issue, that option was shipped long time ago but we didn't force an upgrade to 4.6. And now that we did, that setting does not even work correctly in all scenarios. :(

gabikliot commented:
https://github.com/dotnet/corefx/issues/2454

gabikliot commented:
I remember we asked  the TPL team way back in 2010-2011 for a number of knobs to customize awaiter, to be able to change/override some of those dangerous behaviors with​ Task. Like an ability to override default scheduler, overload Task.Wait, a couple more.
 Unsuccessfully.

jdom commented:
closing by age. Will re-evaluate in the future if it still makes sense to rebase
